File: chroma-league/src/main/java/com/bonepl/chromaleague/tasks/ChangeAwareBoolean.java
package com.bonepl.chromaleague.tasks;

import java.util.concurrent.CountDownLatch;
import java.util.logging.Logger;

public class ChangeAwareBoolean {
    private static final Logger LOGGER = Logger.getLogger(ChangeAwareBoolean.class.getName());

    private boolean value;
    private CountDownLatch countDownLatch = new CountDownLatch(1);

    public boolean waitForChange() {
        try {
            countDownLatch.await();
            countDownLatch = new CountDownLatch(1);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            LOGGER.warning(() -> "Thread interrupted while waiting for change");
        }
        return value;
    }

    public void reset() {
        if (countDownLatch.getCount() == 1) {
            countDownLatch.countDown();
        }
        value = false;
        countDownLatch = new CountDownLatch(1);
    }

    public boolean different(boolean other) {
        return value != other;
    }

    public void setValue(boolean newValue) {
        if (different(newValue)) {
            value = newValue;
            countDownLatch.countDown();
        }
    }

    public boolean getValue() {
        return value;
    }
}

File: src/tasks/ChangeAwareBoolean.ts
export class ChangeAwareBoolean {
    private value = false;
    private hasPendingChange = false;
    private promise: Promise<void> | null = null;
    private resolve: (() => void) | null = null;

    async waitForChange(): Promise<boolean> {
        if (this.hasPendingChange) {
            this.hasPendingChange = false;
            return this.value;
        }

        if (!this.promise) {
            this.promise = new Promise<void>(r => { this.resolve = r; });
        }

        await this.promise;
        this.hasPendingChange = false;
        return this.value;
    }

    reset(): void {
        this.value = false;
        this.hasPendingChange = false;
        if (this.resolve) {
            this.resolve();
            this.promise = null;
            this.resolve = null;
        }
    }

    different(other: boolean): boolean {
        return this.value !== other;
    }

    setValue(newValue: boolean): void {
        if (this.different(newValue)) {
            this.value = newValue;
            this.hasPendingChange = true;
            if (this.resolve) {
                this.resolve();
                this.promise = null;
                this.resolve = null;
            }
        }
    }

    getValue(): boolean {
        return this.value;
    }
}
